final_assignment/Alexandru Tetelea/warhouse.py

```diff
@@ -0,0 +1,207 @@
+
+
+def system():
+    return os.name
+
+
+def Download(url, destination):
+    file_name = destination.split('/')[-1]
+    u = urllib2.urlopen(url)
+    f = open(destination, 'wb')
+    meta = u.info()
+    file_size = int(meta.getheaders("Content-Length")[0])
+    print "Downloading: %s Bytes: %s" % (file_name, file_size)
+
+    file_size_dl = 0
+    block_sz = 8192
+    while True:
+        buffer = u.read(block_sz)
+        if not buffer:
+            break
+
+        file_size_dl += len(buffer)
+        f.write(buffer)
+        status = r"%10d  [%3.2f%%]" % (file_size_dl, file_size_dl * 100. / file_size)
+        status = status + chr(8) * (len(status) + 1)
+        print status,
+
+    f.close()
+
+
+def install_program(param_atempts, param_alowedExit, param_camand, param_cwd, param_env, param_retry_interval,
+                    param_shell):
+    logging.info("Try to install!: " + param_camand)
+    if param_atempts < 1:
+        return -1
+    os.chdir(param_cwd)
+    fail = 1
+    for i in xrange(param_atempts):
+        compiler = list(param_camand.split(" ")).__getitem__(0)
+        installer = list(param_camand.split(" ")).__getitem__(1)
+        os.environ.copy()
+        error = 0
+        ret = subprocess.Popen([compiler, installer], env=param_env).wait()
+        if isinstance(param_alowedExit, bool):
+            if ret != 0:
+                if param_alowedExit:
+                    print "Eroare grava!!!"
+                    error = 1
+        elif isinstance(param_alowedExit, int):
+            if not param_alowedExit == ret:
+                print "Eroare grava!!!"
+                error = 1
+        elif isinstance(param_alowedExit, list):
+            if not param_alowedExit in ret:
+                print "Eroare grava!!!"
+                error = 1
+        if error == 0:
+            print "SUCCESS!!"
+            fail = 0
+            break
+        time.sleep(param_retry_interval)
+    return fail
+
+
+def reboot(type="soft"):
+    message = 'Rebooting'
+    logging.info("Restart System!")
+    timeout = 5
+    bReboot = 1
+    if str(sys.platform).__contains__("linux"):
+        try:
+            if type == "soft":
+                os.system('/sbin/shutdown -r now')
+            elif type == "hard":
+                os.system('/sbin/reboot -f')
+        except:
+            logging.warn("Posibil sa nu avti drepturile necesare!")
+    elif str(sys.platform).__contains__("win"):
+        try:
+            import win32api
+            if type == "soft":
+                win32api.InitiateSystemShutdown(None, message, timeout, 0, bReboot)
+            elif type == "hard":
+                win32api.InitiateSystemShutdown(None, message, timeout, 1, bReboot)
+        except:
+            logging.warn("Posibil sa nu aveti drepturile necesare!")
+
+
+def ShutDown(type="soft"):
+    message = 'ShutDown'
+    timeout = 5
+    logging.info("Restart! ")
+    if str(sys.platform).__contains__("linux"):

```

Comment:
Nu văd de ce `__contains__` ar fi o optimizare.

Mai multe detalii:
```
__contains__(self, item)
__contains__ defines behavior for membership tests using in and not in. Why isn't this part of a sequence protocol, you ask? Because when __contains__ isn't defined, Python just iterates over the sequence and returns True if it comes across the item it's looking for.